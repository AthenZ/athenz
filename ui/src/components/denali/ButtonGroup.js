/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { css, cx } from '@emotion/css';
import { rgba } from 'polished';
import { colors, cssFontStyles } from './styles/index';

const makeCssButtonGroup = (props) => css`
    align-items: stretch;
    background: ${props.dark
        ? rgba(colors.grey300, 0.1)
        : rgba(colors.brand700, 0.05)};
    border-radius: 2px;
    box-sizing: border-box;
    display: inline-grid;
    grid-template-columns: repeat(${props.length}, 1fr);
    height: 28px;
    position: relative;
    label: button-group;

    /* Active box */
    &::before {
        background: ${colors.white};
        border: 2px solid ${colors.blue500};
        border-radius: 2px;
        box-sizing: border-box;
        content: '';
        height: calc(100% + 4px);
        margin: -2px -2px -2px calc(${props.marginRight}% - 2px);
        position: absolute;
        transition: ${!props.noanim ? 'all 0.3s ease-in' : undefined};
        width: calc(${props.segmentPct}% + 4px);
        label: button-group-active-box;
    }
`;

const makeCssButton = (props) => css`
    align-items: center;
    border-radius: 2px;
    box-sizing: border-box;
    color: ${props.active ? colors.black : colors.brand600};
    cursor: ${props.active ? 'initial' : 'pointer'};
    display: flex;
    ${cssFontStyles.default};
    justify-content: center;
    margin: 3px;
    padding: 0 12px;
    transition: ${!props.noanim ? 'color 0.2s ease-in 0.1s' : undefined};
    white-space: nowrap;
    z-index: 1;
    label: button-group-button;
    &:hover {
        color: ${props.active ? colors.black : colors.brand700};
    }
`;

class ButtonGroup extends React.PureComponent {
    render() {
        const { buttons, className, dark, id, noanim, selectedName } =
            this.props;

        if (!buttons.length) return null;

        const selected = selectedName || buttons[0].name;
        const activeIdx = buttons.findIndex((b) => b.name === selected);

        const renderedButtons = buttons.map((button) => {
            const active = selected === button.name;
            const cssProps = { active, noanim };
            const cssButton = makeCssButton(cssProps);
            return (
                <div
                    className={cssButton}
                    data-active={active}
                    id={button.id ? `button-${button.id}` : undefined}
                    key={button.name}
                    onClick={() => !active && this.props.onClick(button)}
                >
                    {button.label}
                </div>
            );
        });

        const length = this.props.buttons.length;
        const segmentPct = 100 / length;
        const marginRight = activeIdx * segmentPct;
        const cssProps = {
            dark,
            length,
            noanim,
            segmentPct,
            marginRight,
        };

        const cssButtonGroup = makeCssButtonGroup(cssProps);
        const classes = cx(cssButtonGroup, 'denali-button-group', className);

        return (
            <div className={classes} id={id} data-testid='buttongroup'>
                {renderedButtons}
            </div>
        );
    }
}

ButtonGroup.propTypes = {
    /** Array of button configs */
    buttons: PropTypes.arrayOf(
        PropTypes.shape({
            name: PropTypes.string.isRequired,
            label: PropTypes.string.isRequired,
            id: PropTypes.string,
            /** Additional data can be placed here that will be returned on `onClick()` */
            extra: PropTypes.object,
        })
    ).isRequired,
    /** Additonal class to apply to the outer div */
    className: PropTypes.string,
    /** Dark color scheme (can be used in conjunction with `primary`/`secondary`) */
    dark: PropTypes.bool,
    /** id to attach to outer div */
    id: PropTypes.string,
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Handler for button click. The clicked button config is returned */
    onClick: PropTypes.func.isRequired,
    /** Specifiy selected button. Defaults to the first button */
    selectedName: PropTypes.string,
};

ButtonGroup.defaultProps = {
    dark: false,
    noanim: false,
};

export default ButtonGroup;
