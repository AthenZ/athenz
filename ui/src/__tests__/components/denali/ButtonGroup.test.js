/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import ButtonGroup from '../../../components/denali/ButtonGroup';

describe('ButtonGroup', () => {
    const onClick = jest.fn();

    afterEach(() => onClick.mockClear());

    const propButtons = [
        { name: 'a', label: 'Button A' },
        { id: 'b', name: 'b', label: 'Button B' },
        { id: 'c', name: 'c', label: 'Button C' },
        { name: 'd', label: 'Button D' },
    ];

    it('should render a button group', () => {
        const { getByTestId } = render(
            <ButtonGroup
                buttons={propButtons}
                selectedName={'b'}
                onClick={onClick}
            />
        );
        const buttongroup = getByTestId('buttongroup');
        const buttonNodes = buttongroup.querySelectorAll('div');

        expect(buttongroup).toMatchSnapshot();
        expect(buttonNodes.length).toBe(4);
        expect(buttonNodes[0]).toHaveAttribute('data-active', 'false');
        expect(buttonNodes[1]).toHaveAttribute('data-active', 'true');
        expect(buttonNodes[2]).toHaveAttribute('data-active', 'false');
        expect(buttonNodes[3]).toHaveAttribute('data-active', 'false');
    });

    it('should render empty button array', () => {
        const { queryByTestId } = render(
            <ButtonGroup buttons={[]} onClick={onClick} />
        );

        expect(queryByTestId('buttongroup')).toBeNull();
    });

    it('should render a dark button group', () => {
        const { getByTestId } = render(
            <ButtonGroup buttons={propButtons} dark onClick={onClick} />
        );
        const buttongroup = getByTestId('buttongroup');

        expect(buttongroup).toMatchSnapshot();
    });

    it('should disable animations', () => {
        const { getByTestId } = render(
            <ButtonGroup buttons={propButtons} noanim onClick={onClick} />
        );
        const buttongroup = getByTestId('buttongroup');

        expect(buttongroup).toMatchSnapshot();
    });

    it('should render with additional className', () => {
        const { getByTestId } = render(
            <ButtonGroup
                buttons={propButtons}
                dark
                className='custom-class'
                onClick={onClick}
            />
        );
        const buttongroup = getByTestId('buttongroup');

        expect(buttongroup).toHaveClass('denali-button-group');
        expect(buttongroup).toHaveClass('custom-class');
    });

    it('should handle selected change with onClick', () => {
        const { getByTestId } = render(
            <ButtonGroup buttons={propButtons} onClick={onClick} />
        );
        const buttongroup = getByTestId('buttongroup');
        const buttonNodes = buttongroup.querySelectorAll('div');

        fireEvent.click(buttonNodes[0]);
        expect(onClick).not.toHaveBeenCalled();

        onClick.mockClear();
        fireEvent.click(buttonNodes[1]);
        expect(onClick).toHaveBeenLastCalledWith({
            id: 'b',
            label: 'Button B',
            name: 'b',
        });

        onClick.mockClear();
        fireEvent.click(buttonNodes[2]);
        expect(onClick).toHaveBeenLastCalledWith({
            id: 'c',
            label: 'Button C',
            name: 'c',
        });

        onClick.mockClear();
        fireEvent.click(buttonNodes[3]);
        expect(onClick).toHaveBeenLastCalledWith({
            label: 'Button D',
            name: 'd',
        });
    });
});
