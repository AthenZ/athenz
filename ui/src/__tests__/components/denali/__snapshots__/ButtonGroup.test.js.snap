// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should disable animations 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2px;
  box-sizing: border-box;
  color: #303030;
  cursor: initial;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 3px;
  padding: 0 12px;
  white-space: nowrap;
  z-index: 1;
}

.emotion-0:hover {
  color: #303030;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2px;
  box-sizing: border-box;
  color: #3697f2;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 3px;
  padding: 0 12px;
  white-space: nowrap;
  z-index: 1;
}

.emotion-1:hover {
  color: #3570f4;
}

.emotion-4 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background: rgba(53,112,244,0.05);
  border-radius: 2px;
  box-sizing: border-box;
  display: inline-grid;
  grid-template-columns: repeat(4,1fr);
  height: 28px;
  position: relative;
}

.emotion-4::before {
  background: #fff;
  border: 2px solid #0072df;
  border-radius: 2px;
  box-sizing: border-box;
  content: '';
  height: calc(100% + 4px);
  margin: -2px -2px -2px calc(0% - 2px);
  position: absolute;
  width: calc(25% + 4px);
}

<div
  class="emotion-4 denali-button-group"
  data-testid="buttongroup"
>
  <div
    class="emotion-0"
    data-active="true"
  >
    Button A
  </div>
  <div
    class="emotion-1"
    data-active="false"
    id="button-b"
  >
    Button B
  </div>
  <div
    class="emotion-1"
    data-active="false"
    id="button-c"
  >
    Button C
  </div>
  <div
    class="emotion-1"
    data-active="false"
  >
    Button D
  </div>
</div>
`;

exports[`ButtonGroup should render a button group 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2px;
  box-sizing: border-box;
  color: #3697f2;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 3px;
  padding: 0 12px;
  -webkit-transition: color 0.2s ease-in 0.1s;
  transition: color 0.2s ease-in 0.1s;
  white-space: nowrap;
  z-index: 1;
}

.emotion-0:hover {
  color: #3570f4;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2px;
  box-sizing: border-box;
  color: #303030;
  cursor: initial;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 3px;
  padding: 0 12px;
  -webkit-transition: color 0.2s ease-in 0.1s;
  transition: color 0.2s ease-in 0.1s;
  white-space: nowrap;
  z-index: 1;
}

.emotion-1:hover {
  color: #303030;
}

.emotion-4 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background: rgba(53,112,244,0.05);
  border-radius: 2px;
  box-sizing: border-box;
  display: inline-grid;
  grid-template-columns: repeat(4,1fr);
  height: 28px;
  position: relative;
}

.emotion-4::before {
  background: #fff;
  border: 2px solid #0072df;
  border-radius: 2px;
  box-sizing: border-box;
  content: '';
  height: calc(100% + 4px);
  margin: -2px -2px -2px calc(25% - 2px);
  position: absolute;
  -webkit-transition: all 0.3s ease-in;
  transition: all 0.3s ease-in;
  width: calc(25% + 4px);
}

<div
  class="emotion-4 denali-button-group"
  data-testid="buttongroup"
>
  <div
    class="emotion-0"
    data-active="false"
  >
    Button A
  </div>
  <div
    class="emotion-1"
    data-active="true"
    id="button-b"
  >
    Button B
  </div>
  <div
    class="emotion-0"
    data-active="false"
    id="button-c"
  >
    Button C
  </div>
  <div
    class="emotion-0"
    data-active="false"
  >
    Button D
  </div>
</div>
`;

exports[`ButtonGroup should render a dark button group 1`] = `
.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2px;
  box-sizing: border-box;
  color: #3697f2;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 3px;
  padding: 0 12px;
  -webkit-transition: color 0.2s ease-in 0.1s;
  transition: color 0.2s ease-in 0.1s;
  white-space: nowrap;
  z-index: 1;
}

.emotion-1:hover {
  color: #3570f4;
}

.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2px;
  box-sizing: border-box;
  color: #303030;
  cursor: initial;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 3px;
  padding: 0 12px;
  -webkit-transition: color 0.2s ease-in 0.1s;
  transition: color 0.2s ease-in 0.1s;
  white-space: nowrap;
  z-index: 1;
}

.emotion-0:hover {
  color: #303030;
}

.emotion-4 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background: rgba(242,242,242,0.1);
  border-radius: 2px;
  box-sizing: border-box;
  display: inline-grid;
  grid-template-columns: repeat(4,1fr);
  height: 28px;
  position: relative;
}

.emotion-4::before {
  background: #fff;
  border: 2px solid #0072df;
  border-radius: 2px;
  box-sizing: border-box;
  content: '';
  height: calc(100% + 4px);
  margin: -2px -2px -2px calc(0% - 2px);
  position: absolute;
  -webkit-transition: all 0.3s ease-in;
  transition: all 0.3s ease-in;
  width: calc(25% + 4px);
}

<div
  class="emotion-4 denali-button-group"
  data-testid="buttongroup"
>
  <div
    class="emotion-0"
    data-active="true"
  >
    Button A
  </div>
  <div
    class="emotion-1"
    data-active="false"
    id="button-b"
  >
    Button B
  </div>
  <div
    class="emotion-1"
    data-active="false"
    id="button-c"
  >
    Button C
  </div>
  <div
    class="emotion-1"
    data-active="false"
  >
    Button D
  </div>
</div>
`;
